consensus/istanbul/core/commit.go

```diff
@@ -207,62 +218,135 @@ func (c *core) handleCheckedCommitForCurrentSequence(msg *istanbul.Message, comm
-
 	// Add the COMMIT message to current round state
 	if err := c.current.AddCommit(msg); err != nil {
 		logger.Error("Failed to record commit message", "m", msg, "err", err)
 		return err
 	}
-	numberOfCommits := c.current.Commits().Size()
+
+	commits := c.current.Commits()
+	numberOfCommits := commits.Size()
 	minQuorumSize := c.current.ValidatorSet().MinQuorumSize()
 	logger.Trace("Accepted commit for current sequence", "Number of commits", numberOfCommits)
 
 	// Commit the proposal once we have enough COMMIT messages and we are not in the Committed state.
 	//
-	// If we already have a proposal, we may have chance to speed up the consensus process
-	// by committing the proposal without PREPARE messages.
 	// TODO(joshua): Remove state comparisons (or change the cmp function)
 	if numberOfCommits >= minQuorumSize && c.current.State().Cmp(StateCommitted) < 0 {
-		logger.Trace("Got a quorum of commits", "tag", "stateTransition", "commits", c.current.Commits)
-		err := c.commit()
+		proposal := c.current.Proposal()
+		// TODO understand how proposal can be nil.
+		if proposal == nil {
+			return nil
+		}
+
+		// Generate aggregate seal
+		aggregatedSeal, err := c.generateAggregateCommittedSeal()
 		if err != nil {
-			logger.Error("Failed to commit()", "err", err)
-			return err
+			logger.Warn("Initial verificaction of aggregate commit signature failed", "err", err)
+			// Remove any bad committed seals and try again if sufficient commits remain
+			c.removeInvalidCommittedSeals()
+			if c.current.Commits().Size() < c.current.ValidatorSet().MinQuorumSize() {
+				// Wait for more commits
+				return nil
+			}
+			aggregatedSeal, err = c.generateAggregateCommittedSeal()
+		}
+		// If there is still an error then sit this round out, we can't continue.
+		if err != nil {
+			nextRound := new(big.Int).Add(c.current.Round(), common.Big1)
+			logger.Warn("Error on commit, waiting for desired round", "reason", "failed to aggregate commit seals", "err", err, "desired_round", nextRound)
+			c.waitForDesiredRound(nextRound)
+			return nil
+		}
+
+		// Set the epoch aggregate seal if this is the last block of the epoch
+		aggregatedEpochValidatorSetSeal := types.IstanbulEpochValidatorSetSeal{}
+		if istanbul.IsLastBlockOfEpoch(proposal.Number().Uint64(), c.config.Epoch) {
+			epochBitmap, epochAggregate, err := AggregateSeals(
+				c.current.Commits(),
+				func(c *istanbul.CommittedSubject) []byte { return c.EpochValidatorSetSeal },
+			)
+			if err != nil {
+				nextRound := new(big.Int).Add(c.current.Round(), common.Big1)
+				logger.Warn("Error on commit, waiting for desired round", "reason", "getAggregatedSeal", "err", err, "desired_round", nextRound)
+				c.waitForDesiredRound(nextRound)
+				return nil
+			}
+
+			aggregatedEpochValidatorSetSeal.Bitmap = epochBitmap
+			aggregatedEpochValidatorSetSeal.Signature = epochAggregate
 		}
 
-	} else if c.current.GetPrepareOrCommitSize() >= minQuorumSize && c.current.State().Cmp(StatePrepared) < 0 {
```

Comment:
GH appears to be mangling this, but it looks like we no longer consider commit message for this transition, so I think it makes sense to remove it from the commit function.

Reply:
Maybe I was looking at an old version and this isn't here anymore.

Reply:
Yeah probably, this was one of the last changes I made.